File: apps/vscode/src/tools/track/styling/symbolInterval.ts
/**
 * Symbol Interval tool implementation.
 * Sets the time interval for displaying position symbols on track features.
 */

import type { TrackFeature } from '@debrief/schemas';
import { isTrackFeature } from '@debrief/schemas';
import type { MCPToolDefinition } from '../../../types/tool';

export interface SymbolIntervalParams {
  interval?: string;
}

export const toolDefinition: MCPToolDefinition = {
  name: 'symbol-interval',
  description: 'Sets the time interval for displaying position symbols on track features.',
  inputSchema: {
    type: 'object',
    properties: {
      features: { type: 'array', items: { type: 'object' } },
      params: { type: 'object', properties: { interval: { type: 'string', description: 'ISO 8601 duration', 'x-debrief-param-type': 'DurationPreset' } } },
    },
  },
  annotations: {
    'debrief:selectionRequirements': [{ kind: 'TRACK', min: 1 }],
    'debrief:category': 'track/styling',
    'debrief:version': '1.0.0',
    'debrief:outputKind': 'mutation/track/styled',
    'debrief:uiCategory': 'style',
  },
};

export function execute(
  features: TrackFeature[],
  params: SymbolIntervalParams,
): TrackFeature[] {
  const interval = params.interval || 'PT15M';

  const modified: TrackFeature[] = [];

  for (const feature of features) {
    if (!isTrackFeature(feature)) {
      continue;
    }

    // Store symbol_interval as a top-level track property so the
    // PositionSymbolsLayer renderer picks it up via props.symbol_interval.
    // Do NOT set show_symbol=true on default_position_style — the interval
    // mechanism in resolvePositionStyle() selectively enables symbols only
    // at positions that match the interval.
    feature.properties.symbol_interval = interval;

    modified.push(feature);
  }

  if (modified.length === 0) {
    throw new Error('No track features found in input');
  }

  return modified;
}

File: services/calc/debrief_calc/tools/track/styling/symbol_interval.py
"""Symbol Interval tool."""

from __future__ import annotations

from typing import Any

from debrief_calc.models import ContextType, SelectionContext, ToolParameter
from debrief_calc.registry import tool


@tool(
    name="symbol-interval",
    description="Sets the time interval for displaying position symbols on track features.",
    input_kinds=["TRACK"],
    output_kind="mutation/track/styled",
    context_type=ContextType.MULTI,
    parameters=[
        ToolParameter(
            name="interval",
            type="string",
            description="ISO 8601 duration (e.g., 'PT15M' for 15 minutes). Defaults to PT15M.",
            required=False,
            param_type="DurationPreset",
        ),
    ],
)
def symbol_interval(context: SelectionContext, params: dict[str, Any]) -> list[dict[str, Any]]:
    """
    Set the time interval for displaying position symbols on track features.

    Args:
        context: SelectionContext with one or more track features
        params: Parameters dict with 'interval' (ISO 8601 duration, defaults to PT15M)

    Returns:
        List of modified track features with updated symbol interval
    """
    interval = params.get("interval") or "PT15M"

    modified = []
    for feature in context.features:
        props = feature.get("properties", {})
        if props.get("kind") != "TRACK":
            continue

        dps = props.setdefault(
            "default_position_style",
            {
                "show_symbol": True,
                "symbol": "circle",
                "show_label": False,
            },
        )
        dps["show_symbol"] = True
        dps["symbol_interval"] = interval

        modified.append(feature)

    if not modified:
        raise ValueError("No track features found in input")

    return modified
